route/route.go

```diff
@@ -769,11 +769,17 @@ type cacheItem struct {
 // get queries the cached items, returning cache hits that have not expired.
 // Cache missed use the configured getFn to populate the cache.
 func (c *environmentCache) get(key string) (string, error) {
+	// get read lock so that we don't attempt to read from the map
+	// while another routine has a write lock and is actively writing
+	// to the map.
+	c.mutex.RLock()
 	if item, ok := c.items[key]; ok {
 		if time.Now().Before(item.expiresAt) {
+			c.mutex.RUnlock()
 			return item.value, nil
 		}
 	}
+	c.mutex.RUnlock()
```

Comment:
I think @senojj is correct regarding defer'ing `RUnlock` because it's in the same func and it wouldn't activate before `Lock` is called.

An alternative might look like this:

```suggestion
	val := ""
	// get read lock so that we don't attempt to read from the map
	// while another routine has a write lock and is actively writing
	// to the map.
	c.mutex.RLock()
	if item, ok := c.items[key]; ok {
		if time.Now().Before(item.expiresAt) {
			val = item.value
		}
	}
	c.mutex.RUnlock()
	if val != "" {
		return val, nil
	}
```

The above suggestion wouldn't allow empty (`""`) cache items, but not sure that's valid anyway.

Reply:
I like that suggestion, and would additionally suggest that you use the zero value of the string type, as opposed to explicitly initializing it to `""`.
```suggestion
	var val string
	// get read lock so that we don't attempt to read from the map
	// while another routine has a write lock and is actively writing
	// to the map.
	c.mutex.RLock()
	if item, ok := c.items[key]; ok {
		if time.Now().Before(item.expiresAt) {
			val = item.value
		}
	}
	c.mutex.RUnlock()
	if val != "" {
		return val, nil
	}
```

Reply:
Yeah, I think that's a good change 👍🏻